### PR TITLE
Preliminary Play 2.7 support

### PR DESCRIPTION
**Pull request by raboof**
Testing with 2.7.0-M2 now, not thoroughly tested yet though.

### Context
Resolve issue #6362 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by marcospereira** [comment]
@raboof you can see how tests had changed for Play 2.6 here: https://github.com/gradle/gradle/pull/3018.

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/platform/PlayMajorVersion.java [hunk 1]
```diff
@@ -29,7 +29,8 @@
     PLAY_2_3_X("2.3.x", "2.11", "2.10"),
     PLAY_2_4_X("2.4.x", "2.11", "2.10"),
     PLAY_2_5_X("2.5.x", "2.11"),
-    PLAY_2_6_X("2.6.x", "2.12", "2.11");
+    PLAY_2_6_X("2.6.x", "2.12", "2.11"),
+    PLAY_2_7_X("2.7.x", "2.12", "2.11");
```
[the remark this answers is not in the document]
Comment (raboof) [reply subprojects/platform-play/src/main/java/org/gradle/play/internal/platform/PlayMajorVersion.java]: Maybe we should do Scala 2.13 support in a separate PR though?

**Comment by raboof** [comment]
I made some updates, but the integration tests fail because they rely on `StaticRoutesGenerator` which was deprecated in 2.6.0 and now removed in 2.7.0. How should we handle this? Rewrite the tests with `InjectedRoutesGenerator`, or would that fail with older versions of Play?

**Comment by marcospereira** [comment]
> How should we handle this? Rewrite the tests with InjectedRoutesGenerator, or would that fail with older versions of Play?

Hum... `InjectedRoutesGenerator` was [introduced in Play 2.4](https://www.playframework.com/documentation/2.6.x/Migration24#Routing) and Gradle supports Play from 2.3 to 2.6 right now. That means you cannot just change to `InjectedRoutesGenerator` because that will make 2.3 fails (I'm not so sure about how Gradle handles this though).

Is it possible to isolate the tests for 2.7 to force the use of `InjectedRoutesGenerator`? Or better, make it the default for 2.7 even if it is
configured to do the opposite (and log a warning). Another idea is to remove Play 2.3 support since this version of Play is not maintained anymore and then use `InjectedRouterGenerator` as you suggested, @raboof.

@blindpirate @marcphilipp WDYT?

**Comment by raboof** [comment]
The newest version in the play 2.3 series was released back in 2015... feels like it might be OK to drop support for it? Or do we need to go through a deprecation cycle for that first?

**Comment by marcphilipp** [comment]
Removal of 2.3 needs a proper deprecation cycle to be removed with Gradle 6.0.

@raboof Could you please post an example of such a problematic test?

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/platform/PlayMajorVersion.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (renatocaval) [reply subprojects/platform-play/src/main/java/org/gradle/play/internal/platform/PlayMajorVersion.java]: we dropped support for Scala 2.13 for now.

File: subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/PlayPlatformIntegrationTest.groovy [hunk 2]
```diff
@@ -65,10 +65,11 @@ model {
...
         where:
-        platform                        | playVersion | scalaPlatform
-        "play: '2.4.5', scala: '2.11'"  | '2.4.5'     | '2.11'
-        "play: '2.4.11', scala: '2.11'" | '2.4.11'    | '2.11'
-        "play: '2.5.18', scala: '2.11'" | '2.5.18'    | '2.11'
+        platform                          | playVersion | scalaPlatform
+        "play: '2.4.5', scala: '2.11'"    | '2.4.5'     | '2.11'
+        "play: '2.4.11', scala: '2.11'"   | '2.4.11'    | '2.11'
+        "play: '2.5.18', scala: '2.11'"   | '2.5.18'    | '2.11'
+        "play: '2.7.0-M2', scala: '2.12'" | '2.7.0-M2'  | '2.12'
```
Comment (renatocaval) [inline subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/PlayPlatformIntegrationTest.groovy]:
```suggestion
        "play: '2.7.0-RC3', scala: '2.12'" | '2.7.0-RC3'  | '2.12'
```

What about 2.6.20?

File: subprojects/platform-play/src/test/groovy/org/gradle/play/internal/PlayPlatformResolverTest.groovy [hunk 3]
```diff
@@ -74,6 +74,7 @@ class PlayPlatformResolverTest extends Specification {
         "2.4.8"     | "2.11.12"
         "2.5.4"     | "2.11.12"
         "2.6.5"     | "2.12.6"
+        "2.7.0-M2"  | "2.12.6"
```
Comment (renatocaval) [inline subprojects/platform-play/src/test/groovy/org/gradle/play/internal/PlayPlatformResolverTest.groovy]:
```suggestion
        "2.7.0-RC3"  | "2.12.6"
```

File: subprojects/platform-play/src/test/groovy/org/gradle/play/internal/PlayPlatformResolverTest.groovy [hunk 4]
```diff
@@ -105,6 +106,7 @@ class PlayPlatformResolverTest extends Specification {
 
         where:
         playVersion | scalaVersion | message
+        '2.7.0-M2'  | '2.10'       | "Play versions 2.7.x are not compatible with Scala platform 2.10. Compatible Scala platforms are [2.12, 2.11]."
```
Comment (renatocaval) [inline subprojects/platform-play/src/test/groovy/org/gradle/play/internal/PlayPlatformResolverTest.groovy]:
```suggestion
        '2.7.0-RC3'  | '2.10'       | "Play versions 2.7.x are not compatible with Scala platform 2.10. Compatible Scala platforms are [2.12, 2.11]."
```

File: subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy [hunk 5]
```diff
@@ -165,6 +165,7 @@ class PlayDistributionPluginTest extends Specification {
         "2.4.11"    | "play.core.server.ProdServerStart"
         "2.5.18"    | "play.core.server.ProdServerStart"
         "2.6.6"     | "play.core.server.ProdServerStart"
+        "2.7.0-M2"  | "play.core.server.ProdServerStart"
```
Comment (renatocaval) [inline subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy]:
```suggestion
        "2.7.0-RC3"  | "play.core.server.ProdServerStart"
```

File: subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/PlayCoverage.groovy [hunk 6]
```diff
@@ -22,8 +22,8 @@ import org.gradle.util.VersionNumber
 import static org.gradle.api.JavaVersion.current
 
 class PlayCoverage {
-    static final List<VersionNumber> ALL_VERSIONS = ["2.4.11", "2.5.18", DefaultPlayPlatform.DEFAULT_PLAY_VERSION].collect { VersionNumber.parse(it) }
-    static final List<VersionNumber> JDK9_COMPATIBLE_VERSIONS = [DefaultPlayPlatform.DEFAULT_PLAY_VERSION]
+    static final List<VersionNumber> ALL_VERSIONS = ["2.4.11", "2.5.18", DefaultPlayPlatform.DEFAULT_PLAY_VERSION, "2.7.0-M2"].collect { VersionNumber.parse(it) }
```
Comment (renatocaval) [inline subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/PlayCoverage.groovy]:
```suggestion
 static final List<VersionNumber> ALL_VERSIONS = ["2.4.11", "2.5.18", DefaultPlayPlatform.DEFAULT_PLAY_VERSION, "2.7.0-RC3"].collect { VersionNumber.parse(it) }
```

File: subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/PlayCoverage.groovy [hunk 7]
```diff
@@ -22,8 +22,8 @@ import org.gradle.util.VersionNumber
...
+    static final List<VersionNumber> JDK9_COMPATIBLE_VERSIONS = [DefaultPlayPlatform.DEFAULT_PLAY_VERSION, "2.7.0-M2"]
```
Comment (renatocaval) [inline subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/PlayCoverage.groovy]:
```suggestion
static final List<VersionNumber> JDK9_COMPATIBLE_VERSIONS = [DefaultPlayPlatform.DEFAULT_PLAY_VERSION, "2.7.0-RC3"]
```

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/twirl/TwirlCompilerFactory.java [hunk 8]
```diff
@@ -37,6 +37,8 @@ public static VersionedTwirlCompilerAdapter createAdapter(PlayPlatform playPlatf
                 return new TwirlCompilerAdapterV10X("1.1.1", scalaCompatibilityVersion, playTwirlAdapter);
             case PLAY_2_6_X:
                 return new TwirlCompilerAdapterV13X("1.3.13", scalaCompatibilityVersion, playTwirlAdapter);
+            case PLAY_2_7_X:
+                return new TwirlCompilerAdapterV13X("1.4.0-M2", scalaCompatibilityVersion, playTwirlAdapter);
```
Comment (renatocaval) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/twirl/TwirlCompilerFactory.java]:
```suggestion
                return new TwirlCompilerAdapterV13X("1.4.0-RC3", scalaCompatibilityVersion, playTwirlAdapter);
```

File: subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/PlayPlatformIntegrationTest.groovy (hunk 2 above)
Comment (raboof) [reply subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/PlayPlatformIntegrationTest.groovy]: Other PR perhaps?

File: subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy [hunk 9]
```diff
@@ -165,6 +165,7 @@ class PlayDistributionPluginTest extends Specification {
         "2.4.11"    | "play.core.server.ProdServerStart"
         "2.5.18"    | "play.core.server.ProdServerStart"
         "2.6.6"     | "play.core.server.ProdServerStart"
+        "2.7.0-RC3"  | "play.core.server.ProdServerStart"
```
Comment (marcospereira) [inline subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayDistributionPluginTest.groovy]:
```suggestion
        "2.7.0"  | "play.core.server.ProdServerStart"
```

File: subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/PlayCoverage.groovy [hunk 10]
```diff
@@ -22,8 +22,8 @@ import org.gradle.util.VersionNumber
 import static org.gradle.api.JavaVersion.current
 
 class PlayCoverage {
-    static final List<VersionNumber> ALL_VERSIONS = ["2.4.11", "2.5.18", DefaultPlayPlatform.DEFAULT_PLAY_VERSION].collect { VersionNumber.parse(it) }
-    static final List<VersionNumber> JDK9_COMPATIBLE_VERSIONS = [DefaultPlayPlatform.DEFAULT_PLAY_VERSION]
+ static final List<VersionNumber> ALL_VERSIONS = ["2.4.11", "2.5.18", DefaultPlayPlatform.DEFAULT_PLAY_VERSION, "2.7.0"].collect { VersionNumber.parse(it) }
```
Comment (marcospereira) [inline subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/PlayCoverage.groovy]: Code format got a little weird here.

**Comment by marcospereira** [comment]
@marcphilipp @blindpirate,

I think this is now complete and good to review/merge. It is not removing Play 2.3 anymore so that it does not need to go through the deprecation cycle.

WDYT?

**Comment by adcalder7** [comment]
Please merge.

**Comment by renatocaval** [comment]
@big-guy / @marcphilipp, is there anything missing to get this merged?

**Comment by big-guy** [comment]
@renatocaval We want to get this in, but we're also trying to extract the Play plugin into a separately releasable plugin: https://github.com/gradle/playframework  We're not going to get to this PR for the next release (5.4), but I want to circle back on this for the following release.

**Comment by dwijnand** [comment]
From https://github.com/gradle/gradle/issues/6362#issuecomment-492107068 it seems the new plan is to migrate/port, not merge this PR.

**Comment by lacasseio** [comment]
We may merge this as-is but the expectation should be to migrate to [gradle/playframework](https://github.com/gradle/playframework). At the moment, this plugin is simply the extraction of the play plugin from the Gradle core to use the current model instead of the software model.

**Comment by arjunchhabra** [comment]
@lacasseio - Any update on this item? Whether extracting or merging here, I believe many of us are currently without a path forward.

**Comment by big-guy** [comment]
I'm going to close this as the playframework plugin has these changes now and should be released shortly.

**Pull request closed without merging**